adr/0010-release-process.md

```diff
@@ -0,0 +1,25 @@
+# Release Process
+
+Date: 2018-12-03
+
+## Status
+
+Draft
+
+## Context
+
+eLife has some projects that are published as packages, e.g. for Python projects on pypi.
+
+Specific details may vary from project to project. For example one project may just pushed to pypi, another to Docker Hub, or even both. Some artefacts may just be pushed for releases, others on commit.
+
+The Build Infrastructure will need to support the release process, including managing the credentials to push the artefact. For that the Build Infrastructure will also need to know when a release should be pushed.
+
+## Decision
+
+Each releasable project will have a mechanism to report on its current version. A release should be pushed, if there isn't currently a tag for the corresponding version (and it's on the master branch). i.e. the version in the develop branch will contain the old version until a new release version should be pushed.
```

Comment:
Keeping the old version until the release seems to be relatively simple. Although semantically it seems counter intuitive:
a) Checking the version of a particular commit would include a release version without indication that it's something between
b) The code may already have breaking changes which are not reflected in the version. To me it seems the major version should change when we make breaking changes, not at release time.

(But then I don't want to engineer a complicated release process either)

Reply:
It may be more complicated yes. If you keep the old version around on an unstable `master` a package generated from that is indistinguishable from the stable version. In the PHP world I've seen people avoiding the version into the code, and this also helps having more than one release from the same commit e.g. `7.2.0-rc6` of PHP becomes the first stable release `7.2.0`. Needs more examples from the open source world.

Reply:
I'd be happy not to include the version in code, although many examples seem to have it. If it's not in the repo, then I guess the release would be manually triggered? (which seem to make sense for releases) That could be done via a parameterised Jenkins task or potentially via the GitHUb release mechanism.

Reply:
Talked in person today, seems to agree on:
- do not include a `__version__` field in the code to avoid the aforementioned problems on multiple packages with the same number
- use `git tag` to indicate a new release, fairly standard in open source
- Jenkins or other systems like Travis CI are able to trigger on the tag in some way to not just run tests but also build a package, push an image, etc (implementation detail)